### PR TITLE
No pushing filters below projections that cast to a lower logical type id

### LINKED ISSUE #12577: Illegal Predicate Pushdown leads to Casting error
### What happens?

The query plan for the query mentioned below rearranges the predicate `WHERE position('*' in b) = 0` such that it is executed *after* the cast which leads to an error. Since predicate and cast specifically happen in the cte, this should not be rearranged.

Plan:
```
┌─────────────────────────────┐
│┌───────────────────────────┐│
││       Physical Plan       ││
│└───────────────────────────┘│
└─────────────────────────────┘
┌───────────────────────────┐
│         PROJECTION        │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│     CAST(a AS INTEGER)    │
│     CAST(b AS INTEGER)    │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│           FILTER          │
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│ ((CAST(col0 AS INTEGER) < │
│ CAST(col1 AS INTEGER)...  │
│ (position(col1, '*') = 0))│
│   ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   │
│           EC: 5           │
└─────────────┬─────────────┘                             
┌─────────────┴─────────────┐
│      COLUMN_DATA_SCAN     │
└───────────────────────────┘  
```

### To Reproduce

Query using CTE:
```sql
WITH t(a, b) AS (
  SELECT a :: int, b :: int
  FROM (VALUES
    ('1', '4'),
    ('5', '3'),
    ('2', '*'),
    ('3', '8'),
    ('7', '*')) AS _(a, b)
  WHERE position('*' in b) = 0
)
SELECT a, b
FROM   t
WHERE  a < b;
```

Same applies when directly inlining:
```sql
SELECT a, b
FROM   (SELECT a :: int, b :: int
  FROM (VALUES
    ('1', '4'),
    ('5', '3'),
    ('2', '*'),
    ('3', '8'),
    ('7', '*')) AS _(a, b)
  WHERE position('*' in b) = 0) AS t(a,b)
WHERE  a < b;
```

### OS:

MacOS 14.3.1: aarch64 (Apple M1 Pro)

### DuckDB Version:

v1.0.1

### DuckDB Client:

CLI

### Full Name:

Louisa Lambrecht

### Affiliation:

University of Tübingen

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

### PR DESCRIPTION
**Pull request by Tmonster**
Fixes https://github.com/duckdb/duckdb/issues/12577

It was also important to realize that if the cast is to a higher logical type, than the filter can be pushed down, since all values of the lower logical type can always be cast to the higher logical type (i.e all INT values can be cast to VARCHAR values).

The other way around, however, does not work, and when such a cast occurs (i.e VARCHAR to INT) the filter cannot be pushed down.

### CODE REVIEW

**Comment by Mytherin** [comment]
Thanks for the PR! Good to consider errors. Another area where we don't consider this right now is the filter re-ordering.

Instead of adding this specifically to the casts specifically in a single optimizer, perhaps we can add this to the `Expression` class, e.g. something like:

```cpp
bool Expression::CanThrow() const override;
```

Then we also need to annotate e.g. scalar functions with whether or not they can throw exceptions. This allows us to solve this problem in a more generic manner.

File: src/include/duckdb/parser/base_expression.hpp [hunk 1]
```diff
@@ -62,6 +62,9 @@ class BaseExpression {
 	virtual bool IsScalar() const = 0;
 	//! Returns true if the expression has a parameter
 	virtual bool HasParameter() const = 0;
+	//! Returns true if executing the base expression can throw an error
+	//! I.e Cast('a' to INT).
+	virtual bool CanThrow() const = 0;
```
Comment (Mytherin) [inline src/include/duckdb/parser/base_expression.hpp:67]: Let's keep this only in the bound expression

**Comment by Tmonster** [comment]
@Mytherin I think the regression test is a false positive. Can you run it again?

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**